### PR TITLE
Add NullableTypeDeclarationForDefaultNullValueFixer

### LINKED ISSUE #3883: Add a rule to force-add or force-remove `?` on nullable types
Sorry if this already exists, I didn't find it with a quick lookup,

Since PHP 7.1, nullable arguments can be written with or without the `?` prefix:
Form 1: `function (string $s = null) {}`
vs
Form 2: `function (?string $s = null) {}`

It would be great to have a rule to normalize to either form 1 or 2.
(On Symfony, we would stick to Form 1 for consistency with branches that still support PHP 5.5)

### PR DESCRIPTION
**Pull request by HypeMC**
Resolves #3883

### CODE REVIEW

File: src/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixer.php [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+            }
+
+            $prevMeaningfulTokenIndex = $tokens->getPrevMeaningfulToken($iter);
+
+            // Check whether the parameter has a type hint
+            if (!$tokens[$prevMeaningfulTokenIndex]->isGivenKind([CT::T_ARRAY_TYPEHINT, T_CALLABLE, T_STRING])) {
+                continue;
+            }
+
+            // Class namespaces are tokenized into multiple tokes, we need to reach the first one
```
Comment (phansys) [inline src/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixer.php]:
```suggestion
            // Class namespaces are tokenized into multiple tokens, we need to reach the first one
```

Comment (HypeMC) [reply src/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixer.php]: @phansys Done

File: tests/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixerTest.php [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PhpCsFixer\Tests\Fixer\FunctionNotation;
+
+use PhpCsFixer\Tests\Test\AbstractFixerTestCase;
+
+/**
+ * @requires PHP 7.1
```
Comment (julienfalque) [inline tests/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixerTest.php]: The tests should run on PHP lower than 7.1 to show that none of the test cases are changed on PHP 5.6 or 7.0, including cases that are changed on PHP 7.1+.

File: src/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixer.php [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+use PhpCsFixer\FixerDefinition\VersionSpecification;
+use PhpCsFixer\FixerDefinition\VersionSpecificCodeSample;
+use PhpCsFixer\Tokenizer\CT;
+use PhpCsFixer\Tokenizer\Token;
+use PhpCsFixer\Tokenizer\Tokens;
+
+/**
+ * @author HypeMC
+ */
+final class NullableTypeHintForDefaultNullValueFixer extends AbstractFixer implements ConfigurationDefinitionFixerInterface
```
Comment (julienfalque) [inline src/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixer.php]: I would rename the class to `NullableTypeDeclarationForDefaultNullValueFixer`.

File: tests/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixerTest.php [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+        yield ['<?php function foo(?\Bar\Baz & $param = null) {}'];
+        yield ['<?php function foo(?array &$param = null) {}'];
+        yield ['<?php function foo(?array & $param = null) {}'];
+        yield ['<?php function foo(?callable &$param = null) {}'];
+        yield ['<?php function foo(?callable & $param = null) {}'];
+        yield ['<?php $foo = function (?Baz $param = null) {};'];
+        yield ['<?php $foo = function (?Baz &$param = null) {};'];
+        yield ['<?php $foo = function (?Baz & $param = null) {};'];
+        yield ['<?php class Test { public function foo(?Bar\Baz $param = null) {} }'];
+        yield ['<?php class Test { public function foo(?self $param = null) {} }'];
```
Comment (julienfalque) [inline tests/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixerTest.php]: Please add test cases with comments and/or whitespaces between `?` and the type.

File: tests/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixerTest.php [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+            '<?php function foo(?iterable $param = null) {}',
+        ];
+        yield [
+            '<?php $foo = function(array $a = null,
+                    array $b = null, array     $c = null, array
+                    $d = null) {};',
+            '<?php $foo = function(?array $a = null,
+                    ?array $b = null, ?array     $c = null, ?array
+                    $d = null) {};',
+        ];
```
Comment (julienfalque) [inline tests/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixerTest.php]: Please add test cases with comments and/or whitespaces between `?` and the type.

File: src/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixer.php [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        return -1;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function createConfigurationDefinition()
+    {
+        return new FixerConfigurationResolver([
+            (new FixerOptionBuilder('use_nullable_type_hint', 'Whether to add or remove `?` before type hints for parameters with a default `null` value.'))
```
Comment (julienfalque) [inline src/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixer.php]: I would rename the option to `use_nullable_type_declaration`.

In the description, _type hints_ should be _type declarations_.

File: src/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixer.php [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+            $prevMeaningfulTokenIndex = $tokens->getPrevMeaningfulToken($iter);
+
+            // We've reached the opening braces so we're done
+            if ($prevMeaningfulTokenIndex <= $startParenthesisIndex) {
+                break;
+            }
+
+            // If the parameter doesn't have a default value null we can continue
+            if (
+                !$tokens[$tokens->getNextMeaningfulToken($variableTokenIndex + 1)]->equals('=')
```
Comment (julienfalque) [inline src/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixer.php]: I think adding 1 to the index will fail if there is no whitespace between the variable token and `=`. Same with adding 2 to find `null` in the next condition (e.g. `string $foo=null`). Please add test cases for those scenarios.

File: tests/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixerTest.php (hunk 2 above)
Comment (HypeMC) [reply tests/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixerTest.php]: @julienfalque Nullable type hints are not supported on lower php versions, a parse error is thrown https://3v4l.org/jPGvV .

Comment (julienfalque) [reply tests/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixerTest.php]: Yes, test cases that remove the `?` should not run on PHP 7.1. However test cases that add it on PHP 7.1+ should run on PHP 5.6 and 7.0 but should expect no change.

**Comment by HypeMC** [comment]
@julienfalque All the requested changes have been made (except the one with the tests being run on older php versions) & can be viewed in this commit https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4401/commits/7143e6d009e2c316e0d376a017269eef15401dbe.

Since the fixer's `isCandidate` method has the following condition, `\PHP_VERSION_ID >= 70100`, running the tests on older php versions results in the test failing with the messegae `Fixer must be a candidate for input code.`, eg https://travis-ci.org/FriendsOfPHP/PHP-CS-Fixer/jobs/563275126. I don't see how the test can be run on older php versions, any help would be appreciated.

File: src/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixer.php [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+ */
+final class NullableTypeHintForDefaultNullValueFixer extends AbstractFixer implements ConfigurationDefinitionFixerInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function getDefinition()
+    {
+        return new FixerDefinition(
+            'Adds or removes `?` before type hints for parameters with a default `null` value.',
```
[the remark this answers is not in the document]
Comment (HypeMC) [reply src/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixer.php]: Done

File: src/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixer.php (hunk 3 above)
Comment (HypeMC) [reply src/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixer.php]: Done

File: tests/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixerTest.php (hunk 4 above)
Comment (HypeMC) [reply tests/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixerTest.php]: Done

File: tests/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixerTest.php (hunk 5 above)
Comment (HypeMC) [reply tests/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixerTest.php]: Done

File: src/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixer.php (hunk 6 above)
Comment (HypeMC) [reply src/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixer.php]: Done

File: src/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixer.php (hunk 7 above)
Comment (HypeMC) [reply src/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixer.php]: Yep, you were right,these cases did fail, fixed now.

File: tests/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixerTest.php (hunk 2 above)
Comment (HypeMC) [reply tests/Fixer/FunctionNotation/NullableTypeHintForDefaultNullValueFixerTest.php]: As mentioned below, not sure how to do this one.

**Comment by julienfalque** [comment]
For test cases that expect no change, the arguments passed to `doTest()` method change: parameter 1 is the input _and_ output code, and parameter 2 is null.

The error you get means parameter 2 is not null and thus treated as input code to be fixed, which requires the `isCandidate()` method to return `true`.

I would either:
- remove the `@requires` annotation on the test method and conditionally pass `null` as parameter 2 to `doTest()` (when PHP version is lower than 7.1)
- add a specific test that runs on PHP lower than 7.1 and always pass `null` as parameter 2 to `doTest()`.

**Comment by julienfalque** [comment]
@HypeMC looking good :+1: 

As this is a new feature, could you please edit the PR to target branch `master` and rebase your work on it?

**Comment by HypeMC** [comment]
@julienfalque Done.

**Comment by SpacePossum** [comment]
Nice MR, great idea :)

The analyzing of the arguments in the fixer is very close to what the [ArgumentsAnalyzer](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/v2.15.1/src/Tokenizer/Analyzer/ArgumentsAnalyzer.php) does, could we utilize that one (maybe together with the FunctionAnalyzer if that helps)?

**Comment by HypeMC** [comment]
@SpacePossum Done, I've used the `FunctionAnalyzer`.

**Comment by drupol** [comment]
@HypeMC Rebase against latest master ?

**Comment by HypeMC** [comment]
@drupol Done.

**Comment by drupol** [comment]
Good, nice feature :-)

**Comment by HypeMC** [comment]
@kubawerlos Tnx for the test, however, there seem to be some problems with the priority. The `NoUnreachableDefaultArgumentValueFixer` has a higher priority the `NullableTypeDeclarationForDefaultNullValueFixer`, but that priority is needed cause of the `CompactNullableTypehintFixer`. Not sure how to go about this one.

**Comment by HypeMC** [comment]
@kubawerlos The only way I was able to make this work was by removing the dependency to the `CompactNullableTypehintFixer` and adding the white space removal to this fixer instead (as you suggested) https://github.com/HypeMC/PHP-CS-Fixer/pull/2

**Comment by HypeMC** [comment]
@kubawerlos Any news on this one, are you ok with the changes suggested in https://github.com/HypeMC/PHP-CS-Fixer/pull/2 ? :smile:

**Comment by kubawerlos** [comment]
@HypeMC please do not `push --force` as it hides what is new and what changes since one gave approve. The person who will do the merge will squash commits anyway.

**Comment by SpacePossum** [comment]
Thank you @HypeMC.

**Pull request merged**